code/harness/Tester.cpp

```diff
@@ -529,102 +643,141 @@ int Tester::test_check_random_permutations(const int num_rounds,
 
-    // Begin fsck timing.
-    time_point<steady_clock> fsck_start_time = steady_clock::now();
+  if (current_test_suite_->GetReorderingCompleted() < num_rounds) {
+    cout << "=============== Unable to find new unique state, stopping at " <<
+      current_test_suite_->GetReorderingCompleted() <<
+      " tests ===============" << endl << endl;
+    log << "=============== Unable to find new unique state, stopping at " <<
+      current_test_suite_->GetReorderingCompleted() <<
+      " tests ===============" << endl << endl;
+  }
+  return SUCCESS;
+}
 
-    // Use popen so that we can throw all the output from fsck into the log that
-    // we are keeping. This information will go just before the summary of what
-    // went wrong in the test.
-    FILE *pipe = popen(command.c_str(), "r");
-    char tmp[128];
-    if (!pipe) {
-      test_info.fs_test.SetError(FileSystemTestResult::kOther);
-      test_info.fs_test.error_description = "error running fsck";
-      test_info.PrintResults(log);
-      test_suite.TallyResult(test_info);
-      continue;
+/*
+ * Replays the operations in the recorded workload, stopping at each Checkpoint
+ * found in the workload. At each Checkpoint, the user test case is called so
+ * that it can check that things are in order.
+ *
+ * TODO(ashmrtn): Should this call a separate method in the user test case or
+ * should it still call the check_test() method?
+ */
+int Tester::test_check_log_replay(std::ofstream& log) {
+  assert(current_test_suite_ != NULL);
+
+  // A single entry in the log data would just be the leading Checkpoint in the
+  // log. In this case, there are no tests to run, so just return.
+  if (log_data.size() == 1) {
+    return SUCCESS;
+  }
+
+  // Skip the first disk write as it is just the Checkpoint at the start of the
+  // log.
+  auto log_iter = log_data.begin() + 1;
+  vector<unsigned int> crash_state;
+  unsigned int crash_state_counter = 1;
+
+  while (log_iter != log_data.end()) {
+    // Keep going through the workload data log until we reach a Checkpoint.
+    // Also, skip the very first checkpoint which occurs at the very beginning
+    // of the log.
+    while (log_iter != log_data.end() &&
+        (!log_iter->is_checkpoint() || log_iter == log_data.begin())) {
+      crash_state.push_back(crash_state_counter);
+      ++crash_state_counter;
+      ++log_iter;
     }
-    while (!feof(pipe)) {
-      char *r = fgets(tmp, 128, pipe);
-      // NULL can be returned on error.
-      if (r != NULL) {
-        test_info.fs_test.fsck_result += tmp;
-      }
+
+    // Required since this is a nested loop and `break` in the inner loop would
+    // just exit that loop, but still run the tests below (which we should skip
+    // if we don't have a Checkpoint?).
+    //
+    // TODO(ashmrtn): See if we should test that the entire log replay
+    // (i.e. when we reach the end of the log. If so, delete the below check).
+    if (log_iter == log_data.end()) {
```

Comment:
Should we check the file system is consistent when we reach the end of the log, regardless of whether or not it's a checkpoint?

Reply:
Yes, that sounds like a reasonable thing to do, as we expect a consistent state when all the bios are replayed.